common/flatpak-run.c

```diff
@@ -1248,6 +1248,49 @@ flatpak_run_add_dconf_args (FlatpakBwrap *bwrap,
   return TRUE;
 }
 
+static gboolean
+flatpak_run_save_environ (GStrv         run_environ,
+                          char          *dir,
+                          GCancellable  *cancellable,
+                          GError       **error)
+{
+  g_autofree char *string = NULL;
+  gsize size = 0;
+  glnx_autofd int dir_fd = -1;
+
+  if (*run_environ)
```

Comment:
Error handling: Whenever we use a `char **` or equivalent, it's somewhat ambiguous whether `NULL` is allowed.

At each level of the stack, you have two options:

* require `run_environ` to have been supplied, as a matter of API;
* don't require it, and gracefully do ... something? ... if it isn't supplied

At the level of `flatpak_run_save_environ` I think the right answer is obvious: if the environment isn't available then it would make no sense (programming error / undefined behaviour) to call `flatpak_run_save_environ`, therefore it should start with:

```
g_return_val_if_fail (run_environ != NULL, FALSE);
```

and then the rest of the function can assume that `run_environ` is non-NULL (possibly empty, but that isn't an error).

For best robustness, it could also check the other arguments:

```
g_return_val_if_fail (dir != NULL, FALSE);
g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
```

At the level of `flatpak_run_app`, it could potentially make sense for `run_environ` to be either mandatory or optional. I personally think it makes most sense for it to be mandatory, because all callers of `flatpak_run_app` should have no trouble populating it, in which case please make that explicit by adding:

```
g_return_val_if_fail (run_environ != NULL, FALSE);
```

The other option for `flatpak_run_app` would be for this parameter to be mandatory, in which case use:

```
  if (run_environ != NULL &&
    !flatpak_run_save_environ (run_environ,
                                 instance_id_host_private_dir,
                                 cancellable,
                                 error))
    return FALSE;
```

... but I think it makes more sense for a null `run_environ` to be treated as a programming error.

Reply:
A personal rule of mine is to never do input/output parameter checks in static functions because static functions are essentially code extracted from another function for better readability and code reuse. Such checks need to be done only at API level because we have no control over the caller in that case.

In other words, coding errors by external users should be caught at API level, and coding errors in internal implementations should have zero tolerance and result in the most disruptive non-functionality possible so they are identified (by the compiler/static code analysis/tests/the developer/users) and fixed quickly.

The added overhead of these checks, both in runtime but mostly in cognitive complexity, while not large, can end up significant when multiple levels nested or adjacent sets of static functions repeat completely redundant and unnecessary checks.

So personally, I prefer not to add such checks to `flatpak_run_save_environ`, if acceptable.

I will say that I usually also extend this rule to private APIs, because private APIs are essentially static functions extracted to a separate file for better code organization or reuse across compilation units (and then made non-static). However, I can understand that this could be considered more controversial.

With that said, I have a feeling that the original author of `flatpak_run_app` were thinking the same because that function checks only one of its input parameters at the moment, and this was a much later addition. It also seems to be the case in a few other private APIs.

With that in mind, I would prefer that we either leave things as-is in regards to function parameter checks or add additional checks only to `flatpak_run_app`. I will however accept any decision; you are the reviewer after all :sweat_smile:

Reply:
The most important thing here is that it's clear whether a NULL GStrv is a programming error (as it would be in `g_strv_contains()` and `g_strv_length()`) or not (as in `g_strdupv()` and `g_settings_set_strv()`), because GStrv and its relatives like `const char * const *` are types for which common practice is inconsistent.

In public APIs, this would require a `g_return_[val_]if_fail()` and a doc-comment. In private interfaces, precisely how it's made clear is less important: it could be a `g_return_[val_]if_fail()`, or a `g_assert()`, or just a pseudo-doc-comment, as long as it's obvious.

Reply:
Thanks for the clarification.

I added `g_assert` to better communicate that `NULL` is not allowed for `run_environ` in `flatpak_run_save_environ` and `flatpak_run_app`.